In [69]:
from sklearn.metrics import cohen_kappa_score as kappa
import pandas as pd
import numpy as np 
import os
import glob
pd.set_option('display.max_rows', 1500)
filenames=[]
path = r"C:\Users\User\OneDrive - aueb.gr\Επιφάνεια εργασίας\Msc\Practical Data Science\A1\HW2"   ###Insert the path here
dirs = os.listdir( path )
os.chdir(path)
print(os.getcwd())

# This would print all the files and directories
for file in dirs:
    filenames.append(file)
print(filenames)    

C:\Users\User\OneDrive - aueb.gr\Επιφάνεια εργασίας\Msc\Practical Data Science\A1\HW2
['f3662001.xlsx', 'f3662003.xlsx', 'f3662006.xlsx', 'f3662009.xlsx', 'f3662010.xlsx', 'f3662011.xlsx', 'f3662013.xlsx', 'f3662014.xlsx', 'f3662015.xlsx', 'f3662016.xlsx', 'f3662018.xlsx', 'f3662020.xlsx', 'f3662021.xlsx', 'f3662023.xlsx']


In [70]:
coders=pd.DataFrame({'A' : []})    ### initialize dataframe


In [71]:
for file in filenames:                               #  stack the columns
    coders=pd.concat([coders,pd.read_excel(file)])

In [72]:
coders=coders.set_index('Id_verse')        #set index as Id_verse
coders.head(100)
coders

A                                               Text  \
Id_verse                                                           
Z_376     NaN  Ελάτε, σεις οι σκλάβες, πέστε μου την πάσα αλή...   
Z_377     NaN  Που πηγεν η Αντρομάχη φεύγοντας η χιονοβραχιον...   
Z_378     NaN  Μήνα σε κάποια συννυφάδα της, μη σε κουνιάδα ε...   
Z_379_380 NaN  για ανέβη με τις καλοπλέξουδες αρχόντισσες, πο...   
Z_382_387 NaN  Έχτορα, τώρα αφού με πρόσταξες να πω την πάσα ...   
Z_388_389 NaN  Πήγε λοιπόν τρεχάτη κι έφτασε στο καστροτείχι ...   
Z_407_410 NaN  Απ᾿ την ορμή την ίδια σου, άμοιρε, θα βρεις το...   
Z_410_413 NaN  Μ᾿ αν είναι να σε χάσω,\nν᾿ ανοίξει η γη να μπ...   
Z_414_419 NaN  Τον κύρη μου ο Αχιλλέας τον σκότωσεν ο αρχοντο...   
Z_419_420 NaN  και φτελιές του φύτεψαν ολόγυρα οι πανώριες Νε...   
Z_421_422 NaN  Κι είχα κι εφτά αδερφούς που χαιρόμουν στο αρχ...   
Z_423_424 NaN  τους σκότωσε ο Αχιλλέας ο πέρφανος, την ώρα πο...   
Z_425_428 NaN  και τη μητέρα μου, που αφέντευε στης δασωμένης...   
Z_429_430 NaN  Έχτορα, τώρα εσύ πατέρας μου και σεβαστή μου μ...   
Z_431_432 NaN  Αχ έλα τώρα πια, σπλαχνίσου μας και μείνε εδώ ...   
Z_433_434 NaN  Στην άγρια τη συκιά το ασκέρι σου για στήσε δί...   
Z_435_437 NaN  Ν᾿ ανέβουν τρεις φορές δοκίμασαν με διαλεγμένο...   
Z_438_439 NaN  θες κάποιος μάντης τους αρμήνεψε, τις θείες βο...   
Z_441_443 NaN   Κι εγώ όλα τούτα τα στοχάζομαι, καλή μου, αλή...   
Z_444_446 NaN  μήτε το λέει η καρδιά μου, τι έμαθα να 'μαι αν...   
Z_447     NaN  τι εγώ στο νου μου και στα φρένα μου καλά το ξ...   
Z_448_449 NaN  Θα ξημερώσει μέρα κάποτε πού θα χαθεί το κάστρ...   
Z_450_454 NaN  Μα τόσο για των Τρωών δε νοιάζομαι τα πάθη όπο...   
Z_455_458 NaN  όσο για σένα, όταν χαλκάρματος κάποιος Αργίτης...   
Z_459     NaN  και κάποιος πει τυχόν, θωρώντας σε να χύνεις μ...   
Z_460_461 NaN  Γιά κοίτα τη γυναίκα του Έχτορα, που ήταν στη ...   
Z_462_463 NaN  Αυτά θα πει, και τότε μέσα σου ξανά θ᾿ ανάψει ...   
Z_464_465 NaN  Μα κάλλιο να μη ζω, να βρίσκομαι βαθιά στη γη ...   
Z_476_478 NaN  Πατέρα Δία κι εσείς οι επίλοιποι θεοί, και τού...   
Z_479     NaN  κι ένας να πει: "πολύ καλύτερος απ᾿ το γονιό τ...   
Z_480_481 NaN  σα θα γυρίζει από τον πόλεμο με κούρσα αιματωμ...   
Z_486     NaN  Άμοιρη εσύ, μη μου πικραίνεσαι μες στην καρδιά...   
Z_487     NaN  κανείς, αν δεν το στρέγει η μοίρα μου, στον Άδ...   
Z_488_489 NaN  το ξέρω, απ᾿ το γραφτό κανένας μας, κιοτής για...   
Z_490_493 NaN  Μον᾿ τώρα εσύ στο σπίτι πήγαινε και τις δουλει...   
Ω_486_487 NaN  Βάλε στο νου, Αχιλλέα θεόμορφε, τον κύρη το δι...   
Ω_488_489 NaN  Μπορεί κι αυτός απ᾿ τους γειτόνους του να τυρα...   
Ω_490_492 NaN  Μα εκείνος, ζωντανός ακούγοντας πως είσαι, ανα...   
Ω_492_494 NaN  Μα εγώ ο τρισάμοιρος, που αξιώθηκα τους γιους ...   
Ω_495     NaN  Είχα πενήντα γιους, σαν έφτασαν οι Αργίτες εδώ...   
Ω_496_497 NaN  οι δεκαεννιά απ᾿ την ίδια εβγήκανε κοιλιά, του...   
Ω_498_501 NaN  Οι πιο πολλοί απ᾿ τον Άρη εχάθηκαν τον άγριο, ...   
Ω_501_502 NaN  Για χάρη του στα πλοία σας φτάνω τώρα,\nνα τον...   
Ω_503_506 NaN  Έλα, σεβάσου τους αθάνατους, συμπόνεσε και μέν...   
Ω_507_508 NaN  Είπε, και τον καημό του εφούντωσε για το δικό ...   
Ω_509_512 NaN  Μαζί τους έπνιξαν οι θύμησες, τον έναν του αντ...   
Ω_513_517 NaN  Μα σύντας ο Αχιλλέας εχόρτασεν ο αρχοντικός το...   
Ω_518     NaN  Άμοιρε εσύ και που ποτίστηκες πικρά φαρμάκια τ...   
Ω_519_521 NaN  Μονάχος να 'ρθεις πώς το βάσταξες στ᾿ Αργίτικα...   
Ω_522_523 NaN  Μον᾿ έλα, στο θρονί για κάθισε, και τους καημο...   
Ω_524     NaN  όχι, δε βγαίνει τίποτα όφελος απ᾿ το φριχτό το...   
Ω_525_526 NaN  Τέτοια οι θεοί μαθές στους άμοιρους θνητούς έκ...   
Ω_527_528 NaN  Μπροστά στου Δία την πόρτα βρίσκουνται στημένα...   
Ω_529_530 NaN  Κι αν δώσει ο Δίας ο κεραυνόχαρος μαζί απ᾿ τα ...   
Ω_531_533 NaN  Μ᾿ απ᾿ τα κακά μονάχα αν του 'δωσε, τον πνίγει...   
Ω_534_537 NaN  Έτσι οι θεοί απ᾿ τα γεννητάτα του και στον Πηλ...   
Ω_538_539 NaN  Όμως ο Δίας με δίχως βάσανα δεν άφησε και τούτ...   


In [73]:
############## Select rows where Text is not NaN

coders=coders.loc[coders["Text"].notna(),:]

##############  select all the rows except the last 3 that dont have index

coders=coders.iloc[:1344,:]

In [74]:
coders.columns

Index(['A', 'Text', 'Subjectivity', 'Sentiment', 'Primary Emotion',
       'Seocondary Emotions', 'Emotion Primary', 'Emotion Secondary',
       'Secondary Emotion'],
      dtype='object')

In [75]:
############# drop unwanted columns

coders=coders.drop(['A','Seocondary Emotions', 'Emotion Primary', 'Emotion Secondary',
       'Secondary Emotion'], axis=1)


In [76]:
################ i made a list of the id's but its a list of lists 

ID_list=[]
for i in range(len(filenames)):
    ID_list.append([filenames[i][1:8]]*96)
    
    
################# To flatten the list of lists i used a flatten method

def flatten(list_of_lists):
    if len(list_of_lists) == 0:
        return list_of_lists
    if isinstance(list_of_lists[0], list):
        return flatten(list_of_lists[0]) + flatten(list_of_lists[1:])
    return list_of_lists[:1] + flatten(list_of_lists[1:])


######### flatten with the method call and create the new column

ID_list=flatten(ID_list) 
coders['ID'] =ID_list    


In [77]:
########################### Boolean array for iloc

a=np.array([True]*1344)
for i in range(1344):
    if(int(coders.ID[i][6])<5):  #########  if last digit <5 the assigned value is G2
        a[i]=True
    else:
        a[i]=False

        

In [78]:
########################### initialize everything false and change the  <5 values to G2


Group_list=[['G1']*1344]
Group_list=flatten(Group_list)
len(Group_list)
coders.loc[:,['Group']]=Group_list

coders.iloc[a,5]='G2'




In [79]:
########################### make it string to apply strip method and remove spaces before and after m + - and 0

coders.iloc[:,2]= coders.iloc[:,2].astype(str)
coders.iloc[:,2]=coders.iloc[:,2].str.strip()

In [80]:
############################    replace Sentiment


coders.iloc[:,2]=coders.iloc[:,2].replace(["+","-", "m"], [1, -1, 1])
coders.iloc[:,2]=coders.iloc[:,2].replace(["+","_", " m"], [1, -1, 1]) ### there was an "_" value i counted it as "-"
coders.iloc[:,2]=coders.iloc[:,2].replace(["nan"], [np.nan])   ######### the nan values of sentiment became string so
#i reassign them as np.nan to compute the nan values in the next question

In [81]:
############################    3

for col_name in coders.columns:
    print(coders[col_name].isna().sum())


0
8
161
412
0
0


In [82]:
############# Fill Na with 0 because we assume that the annotator didnt know what to feel 

coders=coders.replace([np.nan],[0])
for col_name in coders.columns:
    print(coders[col_name].isna().sum())


0
0
0
0
0
0


In [83]:
############################    4

coders.iloc[:,4]=coders.iloc[:,4].astype(str) #### set original index as string
print(type(np.dtype(coders.ID)),type(coders.ID[0]))  #for pandas object=string

coders=coders.set_index(['Group','ID',coders.index ])      #######       set the multindex
coders=coders.sort_index()                                 #######       neaty index

coders

<class 'numpy.dtype[object_]'> <class 'str'>


Text  \
Group ID      Id_verse                                                       
G1    3662006 Z_376      Ελάτε, σεις οι σκλάβες, πέστε μου την πάσα αλή...   
              Z_377      Που πηγεν η Αντρομάχη φεύγοντας η χιονοβραχιον...   
              Z_378      Μήνα σε κάποια συννυφάδα της, μη σε κουνιάδα ε...   
              Z_379_380  για ανέβη με τις καλοπλέξουδες αρχόντισσες, πο...   
              Z_382_387  Έχτορα, τώρα αφού με πρόσταξες να πω την πάσα ...   
              Z_388_389  Πήγε λοιπόν τρεχάτη κι έφτασε στο καστροτείχι ...   
              Z_407_410  Απ᾿ την ορμή την ίδια σου, άμοιρε, θα βρεις το...   
              Z_410_413  Μ᾿ αν είναι να σε χάσω,\nν᾿ ανοίξει η γη να μπ...   
              Z_414_419  Τον κύρη μου ο Αχιλλέας τον σκότωσεν ο αρχοντο...   
              Z_419_420  και φτελιές του φύτεψαν ολόγυρα οι πανώριες Νε...   
              Z_421_422  Κι είχα κι εφτά αδερφούς που χαιρόμουν στο αρχ...   
              Z_423_424  τους σκότωσε ο Αχιλλέας ο πέρφανος, την ώρα πο...   
              Z_425_428  και τη μητέρα μου, που αφέντευε στης δασωμένης...   
              Z_429_430  Έχτορα, τώρα εσύ πατέρας μου και σεβαστή μου μ...   
              Z_431_432  Αχ έλα τώρα πια, σπλαχνίσου μας και μείνε εδώ ...   
              Z_433_434  Στην άγρια τη συκιά το ασκέρι σου για στήσε δί...   
              Z_435_437  Ν᾿ ανέβουν τρεις φορές δοκίμασαν με διαλεγμένο...   
              Z_438_439  θες κάποιος μάντης τους αρμήνεψε, τις θείες βο...   
              Z_441_443   Κι εγώ όλα τούτα τα στοχάζομαι, καλή μου, αλή...   
              Z_444_446  μήτε το λέει η καρδιά μου, τι έμαθα να 'μαι αν...   
              Z_447      τι εγώ στο νου μου και στα φρένα μου καλά το ξ...   
              Z_448_449  Θα ξημερώσει μέρα κάποτε πού θα χαθεί το κάστρ...   
              Z_450_454  Μα τόσο για των Τρωών δε νοιάζομαι τα πάθη όπο...   
              Z_455_458  όσο για σένα, όταν χαλκάρματος κάποιος Αργίτης...   
              Z_459      και κάποιος πει τυχόν, θωρώντας σε να χύνεις μ...   
              Z_460_461  Γιά κοίτα τη γυναίκα του Έχτορα, που ήταν στη ...   
              Z_462_463  Αυτά θα πει, και τότε μέσα σου ξανά θ᾿ ανάψει ...   
              Z_464_465  Μα κάλλιο να μη ζω, να βρίσκομαι βαθιά στη γη ...   
              Z_476_478  Πατέρα Δία κι εσείς οι επίλοιποι θεοί, και τού...   
              Z_479      κι ένας να πει: "πολύ καλύτερος απ᾿ το γονιό τ...   
              Z_480_481  σα θα γυρίζει από τον πόλεμο με κούρσα αιματωμ...   
              Z_486      Άμοιρη εσύ, μη μου πικραίνεσαι μες στην καρδιά...   
              Z_487      κανείς, αν δεν το στρέγει η μοίρα μου, στον Άδ...   
              Z_488_489  το ξέρω, απ᾿ το γραφτό κανένας μας, κιοτής για...   
              Z_490_493  Μον᾿ τώρα εσύ στο σπίτι πήγαινε και τις δουλει...   
              Ω_486_487  Βάλε στο νου, Αχιλλέα θεόμορφε, τον κύρη το δι...   
              Ω_488_489  Μπορεί κι αυτός απ᾿ τους γειτόνους του να τυρα...   
              Ω_490_492  Μα εκείνος, ζωντανός ακούγοντας πως είσαι, ανα...   
              Ω_492_494  Μα εγώ ο τρισάμοιρος, που αξιώθηκα τους γιους ...   
              Ω_495      Είχα πενήντα γιους, σαν έφτασαν οι Αργίτες εδώ...   
              Ω_496_497  οι δεκαεννιά απ᾿ την ίδια εβγήκανε κοιλιά, του...   
              Ω_498_501  Οι πιο πολλοί απ᾿ τον Άρη εχάθηκαν τον άγριο, ...   
              Ω_501_502  Για χάρη του στα πλοία σας φτάνω τώρα,\nνα τον...   
              Ω_503_506  Έλα, σεβάσου τους αθάνατους, συμπόνεσε και μέν...   
              Ω_507_508  Είπε, και τον καημό του εφούντωσε για το δικό ...   
              Ω_509_512  Μαζί τους έπνιξαν οι θύμησες, τον έναν του αντ...   
              Ω_513_517  Μα σύντας ο Αχιλλέας εχόρτασεν ο αρχοντικός το...   
              Ω_518      Άμοιρε εσύ και που ποτίστηκες πικρά φαρμάκια τ...   
              Ω_519_521  Μονάχος να 'ρθεις πώς το βάσταξες στ᾿ Αργίτικα...   
              Ω_522_523  Μον᾿ έλα, στο θρονί για κάθισε, και τους καημο...   
              

In [84]:
coders.index.names

FrozenList(['Group', 'ID', 'Id_verse'])

For question 5,i computed kappa for every possible annotator with each other (for every group and feeling).This way every kappa is computed twice but the mean isnt affected.Also i avoided computations of annotator's with themselves. 


In [85]:
############################    5   

############### For group 1 and Sentiment 

Ids=["3662006","3662009","3662015","3662016","3662018"]       
a=[]
for IDD in Ids:
    for Id in Ids:
        if Id!=IDD:
            a.append(kappa(coders.loc[("G1",IDD),"Sentiment"].astype('int64').values,
                    coders.loc[("G1",Id),"Sentiment"].astype('int64').values))

np.mean(a)            

0.12750739891131063

In [86]:
############### For group 1 and Subjectivity

Ids=["3662006","3662009","3662015","3662016","3662018"]
a=[]
for IDD in Ids:
    for Id in Ids:
        if Id!=IDD:
            a.append(kappa(coders.loc[("G1",IDD),"Subjectivity"].astype('int64').values,
                    coders.loc[("G1",Id),"Subjectivity"].astype('int64').values))

np.mean(a)     

0.08146248468137547

In [87]:
############### For group 2 and Sentiment

Ids=["3662001","3662003","3662010","3662011","3662013","3662014","3662020","3662021","3662023"]
a=[]
for IDD in Ids:
    for Id in Ids:
        if Id!=IDD:
            a.append(kappa(coders.loc[("G2",IDD),"Sentiment"].astype('int64').values,
                    coders.loc[("G2",Id),"Sentiment"].astype('int64').values))

np.mean(a)     

0.18021926901463725

In [88]:
############### For group 2 and Subjectivity

Ids=["3662001","3662003","3662010","3662011","3662013","3662014","3662020","3662021","3662023"]
a=[]
for IDD in Ids:
    for Id in Ids:
        if Id!=IDD:
            a.append(kappa(coders.loc[("G2",IDD),"Subjectivity"].astype('int64').values,
                    coders.loc[("G2",Id),"Subjectivity"].astype('int64').values))

np.mean(a)     

0.1606969302209645

In [89]:
#  G1 SENTI 0.12750739891131063 
#  G1 SUBJ 0.08146248468137547  
#  G2 SENTI 0.18021926901463725  
#  G2 SUBJ 0.1606969302209645   

In conclusion we observe that kappa means are close to zero,so close to randomness.That is a affected by the random selections of the annotators.Subjection of Group 1 has the closest to randomness.Moreover the readers agree more when predicting the Sentiment.But that is caused by the zero fills for G2 beacuse the G2 has 9 annotators and the NA fills for Sentiment were 412.That said,we conclude that  annotators didnt agree in either case and the agreement was poor in each case.